File: src/my_py_pkg/my_py_pkg/add_two_ints_client.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts
import sys

#create a node with the name 'my_first_node', inside the node we will write the code to publish and subscribe to topics
# this node sits inside the 'my_py_pkg' package
#so a pakage is a collection of nodes or a single node

class AddIntsClientNode(Node):
    def __init__(self):
        super().__init__('add_two_ints_client')

        self.server = self.create_client(AddTwoInts, 'add_two_ints')

    
    def send_request(self, a, b):
        self.get_logger().info('Sending request')
        request = AddTwoInts.Request()
        request.a = a
        request.b = b

        future = self.server.call_async(request)
        future.add_done_callback(self.callback)

    def callback(self, future):
        response = future.result()
        print(response.sum)
# this is the main function that will be called when we run the node    
def main(args=None):
    rclpy.init(args=args)

    if (len(sys.argv) != 3):
        print("Usage: ros2 run my_py_pkg add_ints_client_node <a> <b>")
        sys.exit(1)
    a = int(sys.argv[1])
    b = int(sys.argv[2])
    node = AddIntsClientNode()
    node.send_request(a, b)
    rclpy.spin(node) # this will block the program until the node is shutdown
    rclpy.shutdown() # this will shutdown the node


if __name__ == '__main__':
    main()

File: src/my_cpp_pkg/src/add_two_ints_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"

class AddTwoIntsClient : public rclcpp::Node
{
public:
    AddTwoIntsClient() : Node("add_two_ints_client")
    
    {
        client_ = this->create_client<example_interfaces::srv::AddTwoInts>("add_two_ints");
    }


    void send_request(const  int a, const int b)
        {
            // Wait for the service to be available
            while (!client_->wait_for_service(std::chrono::seconds(1)))
            {
                RCLCPP_INFO(this->get_logger(), "Waiting for service");
            }


            auto request = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
            request->a = a; 
            request->b = b;
            client_->async_send_request(request, std::bind(&AddTwoIntsClient::call_back, this, std::placeholders::_1));
        }
private:
    void call_back(rclcpp::Client<example_interfaces::srv::AddTwoInts>::SharedFuture future)
    {
            auto response = future.get();
            RCLCPP_INFO(this->get_logger(), "SUm %d", (int) response->sum);
    }
    rclcpp::Client<example_interfaces::srv::AddTwoInts>::SharedPtr client_;

};


int main(int argc, char * argv[])
{

    rclcpp::init(argc, argv);

    auto node = std::make_shared<AddTwoIntsClient>();
    int a = std::stoi(argv[1]);
    int b = std::stoi(argv[2]);
    node->send_request(a, b);
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;

}
